src/OpenTelemetryEngine/Traces/FunctionsActivityBuilder.cs

```diff
@@ -0,0 +1,81 @@
+//
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+//
+
+using System.Diagnostics;
+using OpenTelemetryEngine.Constants;
+using OpenTelemetryEngine.ResponseObjects;
+
+namespace OpenTelemetryEngine.Traces
+{ 
+    public class FunctionsActivityBuilder
+    {
+        public static ActivitySource sourceInternal = new ActivitySource("AzureFunctionsInternal");
+        public static ActivitySource source = new ActivitySource(OpenTelemetryModuleConstants.ActivitySourceName);
+
+        private static Dictionary<string, Activity> internalActivitiesByInvocationId = new Dictionary<string, Activity>();
+
+        public static FunctionsActivityResponse StartInternalActivity(string invocationId, string traceParent, string traceState) 
+        {
+            Activity? activity = sourceInternal.StartActivity("InternalActivity");
+
+            if (activity == null) 
+            {
+                Console.WriteLine("WARNING: The InternalActivity was null, logs and spans generated in user code may not link properly to host telemetry");
+                return new FunctionsActivityResponse(null);
+            }
+
+            activity.AddTag("invocationId", invocationId);
+
+            if (ActivityContext.TryParse(traceParent, traceState, true, out ActivityContext activityContext))
+            {
+                activity.SetId(traceParent);
+                activity.SetSpanId(activityContext.SpanId.ToString());
+                activity.SetTraceId(activityContext.TraceId.ToString());
+                activity.SetRootId(activityContext.TraceId.ToString());
+            }
+
+            internalActivitiesByInvocationId.Add(invocationId, activity);
+
+            return new FunctionsActivityResponse(activity);
+        }
+
+        public static void StopInternalActivity(string invocationId)
+        {
+            if (internalActivitiesByInvocationId.Keys.Contains(invocationId)) 
+            {
+                internalActivitiesByInvocationId[invocationId].Stop();
```

Comment:
Yes, good catch. Fixed along with the ConcurrentDictionary change. 
It may be worth discussing having a timer-triggered "cleanup" method that periodically sweeps the dictionary looking for spans that are already closed and removes them from the dictionary. As best I can tell, the Stop-OpenTelemetryInvocationInternal cmdlet will always be called by the worker for every invocation where Start-OpenTelemetryInvocationInternal is used, but in the case that soemthing changes or some unexpected behavior occurs and it isn't, the Span stops with the thread and may be safely removed at that point.